proposals/0570-string-interpolation.rst

```diff
@@ -0,0 +1,314 @@
+        go isStart (x:xs) = (if isStart then id else (',' :)) . interpolatePrec 0 x . go False xs
+
+with instances mostly mirroring ``Show``, except for things like ``Char`` or ``String``, which would not include the quotes.
+
+This proposal adds a new class instead of ``Show`` because ``Show`` is intended to return the Haskell source code used to create the value. A quick search indicates that nothing on Stackage currently defines an ``Interpolate`` type class (alternatively, the class + functions could be named ``Display``, which is only in use by ``rio``).
+
+Could optionally provide a newtype for easy derivation with ``DerivingVia``, or even make ``Interpolate`` one of the built-in class that can be derived (like ``Show``):
+::
+
+  newtype InterpolateFromShow a = InterpolateFromShow a
+
+  instance Interpolate (InterpolateFromShow a) where
+    interpolatePrec = showsPrec
+
+  data Foo = Foo Int Bool
+    deriving (Show)
+    deriving (Interpolate) via InterpolateFromShow Foo
+
+We will always require interpolation to be of the form ``${foo}``, unlike other languages where ``$foo`` is allowed for single-variable interpolation. This simplifies the syntax and avoids the need for escaping bare ``$``. Interpolation may be escaped by escaping the dollar sign; e.g. ``"\${foo}"`` is equivalent to ``['$', '{', 'f', 'o', 'o', '}']``.
+
+Any Haskell expression may go inside the braces, but an unmatched ``}`` will immediately end processing. This restriction simplifies the implementation and isn't a big deal in practice, since one could always break it out into a helper variable.
+
+When ``OverloadedStrings`` is enabled, calls ``fromString`` with the result.
+
+If the `multiline proposal gets accepted <https://github.com/ghc-proposals/ghc-proposals/pull/569>`_, interpolation syntax will also be enabled for multiline syntax, e.g.
+::
+
+  -- "Line 1: 100\nLine 2: 200\n"
+  let s =
+        s"""
+        Line 1: ${x}
+        Line 2: ${y}
+        """
+      x = 100
+      y = 200
+
+Examples
+--------
+
+An ``Interpolate`` instance different from ``Show``:
+::
+
+  data TestFailure = TestFailure { message :: String, file :: FilePath, lineNumber :: Int }
+    deriving (Show)
+
+  instance Interpolate TestFailure where
+    interpolate TestFailure{..} = s"${message} (at ${file}:${lineNumber})"
+
+  let failure = TestFailure "Bad value" "test.hs" 10
+
+  -- "TestFailure { message = \\\"Bad value\\\" file = \\\"test.hs\\\", lineNumber = 10 }"
+  let s1 = show failure
+
+  -- "Bad value (at test.hs:10)"
+  let s2 = interpolate failure
+
+  -- "Got a failure: Bad value (at test.hs:10)"
+  let s3 = s"Got a failure: ${failure}"
+
+Interpolating expressions including braces:
+::
+
+  -- "A record: <Person Bob>"
+  let s = s"A record: ${Person{name = "Bob"}}"
+
+  -- parse error: `['` isn't a valid Haskell expression
+  let s2 = s"A brace: ${['}']}"
+
+Effect and Interactions
+-----------------------
+
+No interactions with existing language features.
```

Comment:
You have documented an interaction with `-XOverloadedStrings` in a previous section?

Reply:
Yes, sorry. Updated